# Cleaning Data

# Intro Duplicates

![green-divider](https://user-images.githubusercontent.com/7065401/52071924-c003ad80-2562-11e9-8297-1c6595f8a7ff.png)

## Quick Summary
## What are major import topics features?

- titanic.Fare.str.replace("$","")
- summer.Athlete_Name = summer.Athlete_Name.str.title()
- summer.Athlete_Name.str.strip()

![purple-divider](https://user-images.githubusercontent.com/7065401/52071927-c1cd7100-2562-11e9-908a-dde91ba14e59.png)

## Detection of Duplicates

In [3]:
import pandas as pd

In [4]:
alphabet = pd.DataFrame(["a","b","c","c","d","e","f","g","g","g"], columns = ["Alphabet"])

In [5]:
alphabet

Alphabet
0        a
1        b
2        c
3        c
4        d
5        e
6        f
7        g
8        g
9        g

In [7]:
alphabet.duplicated(keep = False) # boolean series

0    False
1    False
2     True
3     True
4    False
5    False
6    False
7     True
8     True
9     True
dtype: bool

In [10]:
alphabet[alphabet.duplicated(keep = "first")]

Alphabet
3        c
8        g
9        g

## Titanic Dataset

In [5]:
titanic = pd.read_csv("titanic_imp2.csv")

In [6]:
titanic.head()

Survived  Class  Gender   Age  SipSp  ParCH      Fare Emb Deck
0        0      3    male  22.0      1      0     $7.25   S  NaN
1        1      1  female  38.0      1      0  $71.2833   C    C
2        1      3  female  26.0      0      0    $7.925   S  NaN
3        1      1  female  35.0      1      0     $53.1   S    C
4        0      3    male  35.0      0      0     $8.05   S  NaN

In [7]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 894 entries, 0 to 893
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Survived  894 non-null    object
 1   Class     894 non-null    int64 
 2   Gender    894 non-null    object
 3   Age       758 non-null    object
 4   SipSp     894 non-null    int64 
 5   ParCH     894 non-null    int64 
 6   Fare      894 non-null    object
 7   Emb       892 non-null    object
 8   Deck      203 non-null    object
dtypes: int64(3), object(6)
memory usage: 63.0+ KB


In [8]:
pd.to_numeric(titanic.Fare)

ValueError: Unable to parse string "$7.25" at position 0

Unable to parse string $7.25 at position 0, "$" sign makes problem

In [9]:
titanic.Fare.str.replace("$","")

0         7.25
1      71.2833
2        7.925
3         53.1
4         8.05
        ...   
889       30.0
890       7.75
891       10.5
892       14.4
893     7.8958
Name: Fare, Length: 894, dtype: object

In [10]:
titanic.Fare = titanic.Fare.str.replace("$","")

In [11]:
titanic.Fare.head()

0       7.25
1    71.2833
2      7.925
3       53.1
4       8.05
Name: Fare, dtype: object

![purple-divider](https://user-images.githubusercontent.com/7065401/52071927-c1cd7100-2562-11e9-908a-dde91ba14e59.png)

## Olympic Dataset

In [12]:
summer = pd.read_csv("summer_imp.csv")

In [13]:
summer.head(20)

Year    City      Sport Discipline             Athlete Name Country  \
0   1896  Athens   Aquatics   Swimming           HAJOS, Alfred      HUN   
1   1896  Athens   Aquatics   Swimming        HERSCHMANN, Otto      AUT   
2   1896  Athens   Aquatics   Swimming       DRIVAS, Dimitrios      GRE   
3   1896  Athens   Aquatics   Swimming      Malokinis, Ioannis      GRE   
4   1896  Athens   Aquatics   Swimming      Chasapis, Spiridon      GRE   
5   1896  Athens   Aquatics   Swimming   Chorophas, Efstathios      GRE   
6   1896  Athens   Aquatics   Swimming           Hajos, Alfred      HUN   
7   1896  Athens   Aquatics   Swimming        Andreou, Joannis      GRE   
8   1896  Athens   Aquatics   Swimming   Chorophas, Efstathios      GRE   
9   1896  Athens   Aquatics   Swimming           Neumann, Paul      AUT   
10  1896  Athens   Aquatics   Swimming       Pepanos, Antonios      GRE   
11  1896  Athens  Athletics  Athletics           Lane, Francis      USA   
12  1896  Athens  Athletics  Athletics        Szokolyi, Alajos      HUN   
13  1896  Athens  Athletics  Athletics           BURKE, Thomas      USA   
14  1896  Athens  Athletics  Athletics          HOFMANN, Fritz      GER   
15  1896  Athens  Athletics  Athletics          CURTIS, Thomas      USA   
16  1896  Athens  Athletics  Athletics      GOULDING, Grantley      GBR   
17  1896  Athens  Athletics  Athletics       LERMUSIAUX, Albin      FRA   
18  1896  Athens  Athletics  Athletics            FLACK, Edwin      AUS   
19  1896  Athens  Athletics  Athletics           BLAKE, Arthur      USA   

   Gender                       Event       Medal  
0     Men              100M Freestyle  Gold Medal  
1     Men              100M Freestyle      Silver  
2     Men  100M Freestyle For Sailors      Bronze  
3     Men  100M Freestyle For Sailors  Gold Medal  
4     Men  100M Freestyle For Sailors      Silver  
5     Men             1200M Freestyle      Bronze  
6     Men             1200M Freestyle  Gold Medal  
7     Men             1200M Freestyle      Silver  
8     Men              400M Freestyle      Bronze  
9     Men              400M Freestyle  Gold Medal  
10    Men              400M Freestyle      Silver  
11    Men                        100M      Bronze  
12    Men                        100M      Bronze  
13    Men                        100M        Gold  
14    Men                        100M      Silver  
15    Men                110M Hurdles        Gold  
16    Men                110M Hurdles      Silver  
17    Men                       1500M      Bronze  
18    Men                       1500M        Gold  
19    Men                       1500M      Silver

In [15]:
summer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31170 entries, 0 to 31169
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Year          31170 non-null  int64 
 1   City          31170 non-null  object
 2   Sport         31170 non-null  object
 3   Discipline    31170 non-null  object
 4   Athlete Name  31170 non-null  object
 5   Country       31166 non-null  object
 6   Gender        31170 non-null  object
 7   Event         31170 non-null  object
 8   Medal         31170 non-null  object
dtypes: int64(1), object(8)
memory usage: 2.1+ MB


### change the col name from "Athlete Name" to "Athlete_Name"

In [16]:
summer.Athlete Name

SyntaxError: invalid syntax (<ipython-input-16-52579571f72a>, line 1)

In [17]:
summer.rename( columns = {"Athlete Name" : "Athlete_Name"}, inplace = True)

In [18]:
summer.head(20)

Year    City      Sport Discipline             Athlete_Name Country  \
0   1896  Athens   Aquatics   Swimming           HAJOS, Alfred      HUN   
1   1896  Athens   Aquatics   Swimming        HERSCHMANN, Otto      AUT   
2   1896  Athens   Aquatics   Swimming       DRIVAS, Dimitrios      GRE   
3   1896  Athens   Aquatics   Swimming      Malokinis, Ioannis      GRE   
4   1896  Athens   Aquatics   Swimming      Chasapis, Spiridon      GRE   
5   1896  Athens   Aquatics   Swimming   Chorophas, Efstathios      GRE   
6   1896  Athens   Aquatics   Swimming           Hajos, Alfred      HUN   
7   1896  Athens   Aquatics   Swimming        Andreou, Joannis      GRE   
8   1896  Athens   Aquatics   Swimming   Chorophas, Efstathios      GRE   
9   1896  Athens   Aquatics   Swimming           Neumann, Paul      AUT   
10  1896  Athens   Aquatics   Swimming       Pepanos, Antonios      GRE   
11  1896  Athens  Athletics  Athletics           Lane, Francis      USA   
12  1896  Athens  Athletics  Athletics        Szokolyi, Alajos      HUN   
13  1896  Athens  Athletics  Athletics           BURKE, Thomas      USA   
14  1896  Athens  Athletics  Athletics          HOFMANN, Fritz      GER   
15  1896  Athens  Athletics  Athletics          CURTIS, Thomas      USA   
16  1896  Athens  Athletics  Athletics      GOULDING, Grantley      GBR   
17  1896  Athens  Athletics  Athletics       LERMUSIAUX, Albin      FRA   
18  1896  Athens  Athletics  Athletics            FLACK, Edwin      AUS   
19  1896  Athens  Athletics  Athletics           BLAKE, Arthur      USA   

   Gender                       Event       Medal  
0     Men              100M Freestyle  Gold Medal  
1     Men              100M Freestyle      Silver  
2     Men  100M Freestyle For Sailors      Bronze  
3     Men  100M Freestyle For Sailors  Gold Medal  
4     Men  100M Freestyle For Sailors      Silver  
5     Men             1200M Freestyle      Bronze  
6     Men             1200M Freestyle  Gold Medal  
7     Men             1200M Freestyle      Silver  
8     Men              400M Freestyle      Bronze  
9     Men              400M Freestyle  Gold Medal  
10    Men              400M Freestyle      Silver  
11    Men                        100M      Bronze  
12    Men                        100M      Bronze  
13    Men                        100M        Gold  
14    Men                        100M      Silver  
15    Men                110M Hurdles        Gold  
16    Men                110M Hurdles      Silver  
17    Men                       1500M      Bronze  
18    Men                       1500M        Gold  
19    Men                       1500M      Silver

Some names are in upper_case some of them are lower.

Series.str.title : Converts first character of each word to uppercase and
    remaining to lowercase.

In [19]:
summer.Athlete_Name.str.title()

0                     Hajos, Alfred 
1                  Herschmann, Otto 
2                 Drivas, Dimitrios 
3                Malokinis, Ioannis 
4                Chasapis, Spiridon 
                    ...             
31165            Janikowski, Damian 
31166     Rezaei, Ghasem Gholamreza 
31167                Totrov, Rustam 
31168             Aleksanyan, Artur 
31169                Lidberg, Jimmy 
Name: Athlete_Name, Length: 31170, dtype: object

In [20]:
summer.Athlete_Name = summer.Athlete_Name.str.title()

In [21]:
summer.head(10)

Year    City     Sport Discipline             Athlete_Name Country Gender  \
0  1896  Athens  Aquatics   Swimming           Hajos, Alfred      HUN    Men   
1  1896  Athens  Aquatics   Swimming        Herschmann, Otto      AUT    Men   
2  1896  Athens  Aquatics   Swimming       Drivas, Dimitrios      GRE    Men   
3  1896  Athens  Aquatics   Swimming      Malokinis, Ioannis      GRE    Men   
4  1896  Athens  Aquatics   Swimming      Chasapis, Spiridon      GRE    Men   
5  1896  Athens  Aquatics   Swimming   Chorophas, Efstathios      GRE    Men   
6  1896  Athens  Aquatics   Swimming           Hajos, Alfred      HUN    Men   
7  1896  Athens  Aquatics   Swimming        Andreou, Joannis      GRE    Men   
8  1896  Athens  Aquatics   Swimming   Chorophas, Efstathios      GRE    Men   
9  1896  Athens  Aquatics   Swimming           Neumann, Paul      AUT    Men   

                        Event       Medal  
0              100M Freestyle  Gold Medal  
1              100M Freestyle      Silver  
2  100M Freestyle For Sailors      Bronze  
3  100M Freestyle For Sailors  Gold Medal  
4  100M Freestyle For Sailors      Silver  
5             1200M Freestyle      Bronze  
6             1200M Freestyle  Gold Medal  
7             1200M Freestyle      Silver  
8              400M Freestyle      Bronze  
9              400M Freestyle  Gold Medal

In [22]:
summer.loc[summer.Athlete_Name == "Hajos, Alfred"]

Empty DataFrame
Columns: [Year, City, Sport, Discipline, Athlete_Name, Country, Gender, Event, Medal]
Index: []

- Q: Why we can not see Hajos, Alfred?
- A: is below

In [23]:
summer.iloc[0,4]

' Hajos, Alfred '

There is a empty space in name.

In [24]:
summer.Athlete_Name.str.strip()

0                    Hajos, Alfred
1                 Herschmann, Otto
2                Drivas, Dimitrios
3               Malokinis, Ioannis
4               Chasapis, Spiridon
                   ...            
31165           Janikowski, Damian
31166    Rezaei, Ghasem Gholamreza
31167               Totrov, Rustam
31168            Aleksanyan, Artur
31169               Lidberg, Jimmy
Name: Athlete_Name, Length: 31170, dtype: object

In [25]:
summer.Athlete_Name = summer.Athlete_Name.str.strip()

In [26]:
summer.loc[summer.Athlete_Name == "Hajos, Alfred"]

Year    City     Sport Discipline   Athlete_Name Country Gender  \
0  1896  Athens  Aquatics   Swimming  Hajos, Alfred     HUN    Men   
6  1896  Athens  Aquatics   Swimming  Hajos, Alfred     HUN    Men   

             Event       Medal  
0   100M Freestyle  Gold Medal  
6  1200M Freestyle  Gold Medal

In [27]:
summer.loc[summer.Athlete_Name == "Phelps, Michael"]

Year     City     Sport Discipline     Athlete_Name Country Gender  \
25229  2004   Athens  Aquatics   Swimming  Phelps, Michael     USA    Men   
25257  2004   Athens  Aquatics   Swimming  Phelps, Michael     USA    Men   
25262  2004   Athens  Aquatics   Swimming  Phelps, Michael     USA    Men   
25269  2004   Athens  Aquatics   Swimming  Phelps, Michael     USA    Men   
25281  2004   Athens  Aquatics   Swimming  Phelps, Michael     USA    Men   
25290  2004   Athens  Aquatics   Swimming  Phelps, Michael     USA    Men   
25329  2004   Athens  Aquatics   Swimming  Phelps, Michael     USA    Men   
25365  2004   Athens  Aquatics   Swimming  Phelps, Michael     USA    Men   
27228  2008  Beijing  Aquatics   Swimming  Phelps, Michael     USA    Men   
27256  2008  Beijing  Aquatics   Swimming  Phelps, Michael     USA    Men   
27262  2008  Beijing  Aquatics   Swimming  Phelps, Michael     USA    Men   
27268  2008  Beijing  Aquatics   Swimming  Phelps, Michael     USA    Men   
27280  2008  Beijing  Aquatics   Swimming  Phelps, Michael     USA    Men   
27295  2008  Beijing  Aquatics   Swimming  Phelps, Michael     USA    Men   
27331  2008  Beijing  Aquatics   Swimming  Phelps, Michael     USA    Men   
27370  2008  Beijing  Aquatics   Swimming  Phelps, Michael     USA    Men   
29275  2012   London  Aquatics   Swimming  Phelps, Michael     USA    Men   
29303  2012   London  Aquatics   Swimming  Phelps, Michael     USA    Men   
29314  2012   London  Aquatics   Swimming  Phelps, Michael     USA    Men   
29345  2012   London  Aquatics   Swimming  Phelps, Michael     USA    Men   
29375  2012   London  Aquatics   Swimming  Phelps, Michael     USA    Men   
29410  2012   London  Aquatics   Swimming  Phelps, Michael     USA    Men   

                        Event   Medal  
25229          100M Butterfly    Gold  
25257          200M Butterfly    Gold  
25262          200M Freestyle  Bronze  
25269  200M Individual Medley    Gold  
25281  400M Individual Medley    Gold  
25290  4X100M Freestyle Relay  Bronze  
25329     4X100M Medley Relay    Gold  
25365  4X200M Freestyle Relay    Gold  
27228          100M Butterfly    Gold  
27256          200M Butterfly    Gold  
27262          200M Freestyle    Gold  
27268  200M Individual Medley    Gold  
27280  400M Individual Medley    Gold  
27295  4X100M Freestyle Relay    Gold  
27331     4X100M Medley Relay    Gold  
27370  4X200M Freestyle Relay    Gold  
29275          100M Butterfly    Gold  
29303          200M Butterfly  Silver  
29314             200M Medley    Gold  
29345        4X100M Freestyle  Silver  
29375           4X100M Medley    Gold  
29410        4X200M Freestyle    Gold